Step-by-step-Tutorials/Failure-Handling.md

```diff
@@ -0,0 +1,59 @@
+- Kill all related activations as well to ensure a clean state for all of them.
+- Due multi-grain state manipulations using a [Process Manager](https://msdn.microsoft.com/en-us/library/jj591569.aspx) or database transaction to make sure it's either done completely or nothing is changed to avoid the state being messed up,.
+
+Depending on your application the retry logic might follow a simple or complex pattern and you might have to do other stuff like notifying external systems and other things but generally you either have to retry your action, restart the grain/grains involved or stop responding until something which is not available becomes available. 
+
+If you have a grain responsible for database saves and database is not available, you simply have to fail any request until the DB comes back online. If your operation can be simply be retried at user's will like failure of saving a comment in a comment grain, you can just retry when user presses the retry button (until a certain number of times in order to not saturate the network). Specific details of how to do it are really application specific but the possible strategies are the same.
+
+## Strategy parameters and choosing a good strategy
+
+As described in the section above, choosing a strategy is application and context dependent and strategies usually have parameters which again have to be decided at the application level. For example you might want to retry a request maximum 5 times per minute and can deal with it being done eventually but for some other action you might not be able to continue if something is resolved. If your Login grain fails , you can not process any other requests from the user as an example. 
+
+## A note about supervision trees
+
+Developers used to Erlang/Akka/Akka.Net might be surprised to see that there is no supervision tree in Orleans and wonder how they can recover from failures. The point is that in Orleans since actors are reactivated automatically and you don't handle their life-cycle, you usually only deal with killing actors (deactivating them) and not their creation. If you need to store relations between grains, you can simply do it and it is a widely done practice so if my `GameManager`grain fails, all games belong to it should fail as well and the manager can do this for them by calling `Deactivate OnIdle` 
+
+The difference is, if one of the games receive a call, it will be reactivated automatically so if you need the manager to manage the game grains then all calls to the game should go through the `GameManager` so be careful to honor your hierarchies just like what you do in Erlang/Akka/Akka.Net. If you are making workers for a supervisor process then only the superwiser should communicate with the workers and no one else. In other words if something needs supervision, you need to do the supervision by some grain and Orleans doesn't solve it automagically for you.
```

Comment:
As I'm reading these paragraphs, it seems you are showing how to workaround the lack of supervision trees by emulating (shimming) that missing feature. In my experience using Orleans, you (typically) can get away without supervising grain activations at all, and hence why that doesn't even exist in Orleans. If you model your grains so that they don't depend on the running activations, but instead on whether a particular grain call succeeded or not, then why would you need to make a decision whether the other activation need to be killed or not. You can just assume that other activation virtually exists all the time, it's just that sometimes calls to it fail.
Of course when being are pragmatic, there can be exceptions to this. In that case you *might* consider relying on a specific activation or fail the entire tree (as in Erlang)... If you do this, make sure you are explicitly opting to do so. Again, the note might be useful for this exceptional case, I would just make it clear that it's not how you should typically model your grains, as it will start creating friction with the concept of a *virtual* actor